src/haystack/val/dis_macro.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright (C) 2020 - 2024, J2 Innovations
+
+use std::{borrow::Cow, sync::OnceLock};
+
+use super::Value;
+use regex::{Captures, Match, Regex, Replacer};
+
+/// A regular expression replacer implementation for replacing formatted
+/// values in a display macro string.
+struct DisReplacer<'a, 'b, GetValueFunc, GetLocalizedFunc>
+where
+    GetValueFunc: Fn(&str) -> Option<&'a Value>,
+    GetLocalizedFunc: Fn(&str) -> Option<Cow<'a, str>>,
+{
+    get_value: &'b GetValueFunc,
+
+    get_localized: &'b GetLocalizedFunc,
+}
+
+impl<'a, 'b, GetValueFunc, GetLocalizedFunc> Replacer
+    for DisReplacer<'a, 'b, GetValueFunc, GetLocalizedFunc>
+where
+    GetValueFunc: Fn(&str) -> Option<&'a Value>,
+    GetLocalizedFunc: Fn(&str) -> Option<Cow<'a, str>>,
+{
+    fn replace_append(&mut self, caps: &Captures<'_>, dst: &mut String) {
+        let default_replace = |dst: &mut String| dst.push_str(caps.get(0).unwrap().as_str());
+
+        let mut handle_value_capture = |cap_match: Match<'_>| {
+            if let Some(value) = (self.get_value)(cap_match.as_str()) {
+                if let Value::Ref(val) = value {
+                    dst.push_str(val.dis.as_ref().unwrap_or(&val.value));
+                } else if let Value::Str(val) = value {
+                    dst.push_str(&val.value);
+                } else {
+                    dst.push_str(&value.to_string())
+                }
+            } else {
+                default_replace(dst);
+            }
+        };
+
+        // Replace $tag
+        if let Some(cap_match) = caps.get(2) {
```

Comment:
`caps.get(2)` and `caps.get(4)` could be combined: 
```rust
if let Some(cap_match) = caps.get(2).or_else(|| caps.get(4)) {
    handle_value_capture(cap_match);
}
```

Reply:
I was wondering how I could collapse those checks together. Thanks!